File: RecyclerView/app/src/main/java/com/xyz/recyclerview/MainActivity.java
package com.xyz.recyclerview;

import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private List<Menu> menuList;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        recyclerView = findViewById(R.id.rvMenu);
        buildRecyclerData();
        setRecyclerAdapter();
    }

    private void setRecyclerAdapter() {
        MenuAdapter menuAdapter = new MenuAdapter(menuList);
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        /*
        Pass the data to the adapter class
         */
        recyclerView.setAdapter(menuAdapter);
    }

    private void buildRecyclerData() {
        menuList = new ArrayList<>();
        menuList.add(new Menu("Masala Dosa", 60, "Anand Bhavan"));
        menuList.add(new Menu("Idli Vada", 40, "Anand Bhavan"));
        menuList.add(new Menu("Paneer Masala Dosa", 90, "A2B"));
        menuList.add(new Menu("Masala Dosa", 60, "Anand Bhavan"));
        menuList.add(new Menu("Masala Dosa", 60, "Anand Bhavan"));
        menuList.add(new Menu("Masala Dosa", 60, "Anand Bhavan"));
        menuList.add(new Menu("Masala Dosa", 60, "Anand Bhavan"));
        menuList.add(new Menu("Masala Dosa", 60, "Anand Bhavan"));
        menuList.add(new Menu("Masala Dosa", 60, "Anand Bhavan"));
        menuList.add(new Menu("Masala Dosa", 60, "Anand Bhavan"));
        menuList.add(new Menu("Idli Vada", 40, "Anand Bhavan"));
        menuList.add(new Menu("Paneer Masala Dosa", 90, "A2B"));
        menuList.add(new Menu("Masala Dosa", 60, "Anand Bhavan"));
        menuList.add(new Menu("Masala Dosa", 60, "Anand Bhavan"));
        menuList.add(new Menu("Masala Dosa", 60, "Anand Bhavan"));
        menuList.add(new Menu("Masala Dosa", 60, "Anand Bhavan"));
        menuList.add(new Menu("Masala Dosa", 60, "Anand Bhavan"));
        menuList.add(new Menu("Masala Dosa", 60, "Anand Bhavan"));

        /*
        comment the lines from 35 to 52 and uncomment this for loop
         */
//        for (int i = 0; i < 100;i++){
//            menuList.add(new Menu("Masala Dosa", 60, "Anand Bhavan"));
//        }
    }
}

File: MVVM Architecture/app/src/main/java/com/xyz/mvvmarchitecture/activities/MainActivity.kt
package com.xyz.mvvmarchitecture.activities

import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.xyz.mvvmarchitecture.R
import com.xyz.mvvmarchitecture.adapter.UserAdapter
import com.xyz.mvvmarchitecture.model.DataModel
import com.xyz.mvvmarchitecture.model.UserUIModel
import com.xyz.mvvmarchitecture.viewmodel.UserViewModel
import kotlinx.android.synthetic.main.activity_main.*

/**
 * This Activity makes an Api call and populates the result in a recycler view. This activity has
 * only UI related code
 */
class MainActivity : AppCompatActivity() {

    private lateinit var userViewModel: UserViewModel
    private lateinit var userAdapter: UserAdapter
    private val dataModelList = emptyList<DataModel>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        userViewModel = ViewModelProvider(this).get(UserViewModel::class.java)
        setRecyclerAdapter()
        observeLiveData()
        flProgressBar.visibility = View.VISIBLE
        userViewModel.callAPI()
    }

    /**
     * This method is used to observe the live data changes. i.e it gets called when the Api response is fetched
     */
    private fun observeLiveData() {
        userViewModel.liveData.observe(this, {
            when (it) {
                is UserUIModel.Success -> {
                    userAdapter.updateList(it.dataModelList)
                    flProgressBar.visibility = View.GONE
                }

                is UserUIModel.Failure -> {
                    Toast.makeText(
                        this@MainActivity,
                        "Error message ${it.error}",
                        Toast.LENGTH_SHORT
                    ).show()
                    flProgressBar.visibility = View.GONE
                }
            }
        })
    }

    /**
     * Sets the recycler view adapter
     */
    private fun setRecyclerAdapter() {
        userAdapter = UserAdapter(dataModelList)
        val layoutManager = LinearLayoutManager(this)
        recyclerView.apply {
            this.layoutManager = layoutManager
            adapter = userAdapter
        }
    }
}